File: courseMaterial/course_ROSActions/code_section_2_simple_actions/my_robot_tutorials/src/count_until_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>

#include <my_robot_msgs/CountUntilAction.h>
#include <my_robot_msgs/CountUntilGoal.h>
#include <my_robot_msgs/CountUntilResult.h>
#include <my_robot_msgs/CountUntilFeedback.h>

class CountUntilServer {

	protected:
	ros::NodeHandle _nh;
	actionlib::SimpleActionServer<my_robot_msgs::CountUntilAction> _as;
	int _counter;
	
	public:
	CountUntilServer(): 
		_as(_nh, "/count_until", 
			boost::bind(&CountUntilServer::onGoal, this, _1), 
			false),
		_counter(0)
	{
		_as.start();
		ROS_INFO("Simple Action Server has been started.");
	}

	void onGoal(
		const my_robot_msgs::CountUntilGoalConstPtr &goal) 
	{
		ROS_INFO("Goal received"); 

		int max_number = goal->max_number;
		double wait_duration = goal->wait_duration;
		ROS_INFO("Max number : %d, wait duration : %lf", 
			max_number, wait_duration);

		_counter = 0;
		ros::Rate rate(1.0/wait_duration);

		bool success = false;
		bool preempted = false;

		while (ros::ok()) {
			_counter++;
			if (_as.isPreemptRequested()) {
				preempted = true;
				break;
			}
			if (_counter > 9) {
				break;
			}
			if (_counter > max_number) {
				success = true;
				break;
			}
			ROS_INFO("%d", _counter);
			my_robot_msgs::CountUntilFeedback feedback;
			feedback.percentage = 
				(double)_counter / (double)max_number;
			_as.publishFeedback(feedback);
			rate.sleep();
		}
		

		my_robot_msgs::CountUntilResult result;
		result.count = _counter;
		ROS_INFO("Send goal result to client");

		if (preempted) {
			ROS_INFO("Preempted");
			_as.setPreempted(result);
		}
		else if (success) {
			ROS_INFO("Success");
			_as.setSucceeded(result);
		}
		else {
			ROS_INFO("Aborted");
			_as.setAborted(result);
		}
	}
};

int main (int argc, char **argv) {
	ros::init(argc, argv, "count_until_server");

	ROS_INFO("About to start server");

	CountUntilServer server;
	ros::spin();
}

File: courseMaterial/course_ROSActions/code_section_5_actions_activity/my_robot_tutorials/scripts/count_until_server.py
#!/usr/bin/env python

import rospy
import actionlib

from my_robot_msgs.msg import CountUntilAction
from my_robot_msgs.msg import CountUntilGoal
from my_robot_msgs.msg import CountUntilResult
from my_robot_msgs.msg import CountUntilFeedback

class CountUntilServer:

    def __init__(self):
        self._as = actionlib.SimpleActionServer('/count_until',
                CountUntilAction, execute_cb=self.on_goal, 
				auto_start=False)
        self._as.start()

        self._counter = 0
        rospy.loginfo("Simple Action Server has been started")

    def on_goal(self, goal):
        rospy.loginfo("A goal has been received!")
        rospy.loginfo(goal)

        max_number = goal.max_number
        wait_duration = goal.wait_duration

        self._counter = 0
        rate = rospy.Rate(1.0/wait_duration)

        success = False
        preempted = False

        while not rospy.is_shutdown():
            self._counter += 1
            if self._as.is_preempt_requested():
                preempted = True
                break
            if self._counter > 9:
                break
            if self._counter >= max_number:
                success = True
                break
            rospy.loginfo(self._counter)
            feedback = CountUntilFeedback()
            feedback.percentage = \
				float(self._counter) / float(max_number)
            self._as.publish_feedback(feedback)
            rate.sleep()

        result = CountUntilResult()
        result.count = self._counter
        rospy.loginfo("send goal result to client")

        if preempted:
            rospy.loginfo("Preempted")
            self._as.set_preempted(result)
        elif success:
            rospy.loginfo("Success")
            self._as.set_succeeded(result)
        else:
            rospy.loginfo("Failure")
            self._as.set_aborted(result)

if __name__ == '__main__':
    rospy.init_node('count_until_server')

    server = CountUntilServer()

    rospy.spin()
